File: pl0_scanner.cpp
#include <iostream>
#include <algorithm>
#include <regex>
#include <string>
#include <sstream>

using std::string;
using std::stringstream;
using std::regex;
using std::regex_match;

using std::cin;
using std::cout;

/*
 * c++ 版本的词法分析器
 */

const string digit        = R"([0-9])"

const string keywords     = R"(const|var|procedure|begin|end|if|then"
								"|while|do|read|call|write|writeln|module|int)"

const string constant     = R"([0-9]+)"

const string identfier    = R"([A-Za-z_][A-Za-Z_0-9]*)"

const string operator_tag = R"(\+|-|\*|\/|#|=|<|>|<>|:=|>=|<=)"

const string delimiter    = R"([,.;():])"

const string delim		= R"([ \n\t])"

const string whitespace	= R"([ \n\t]+)"

enum word_type
{
	digit_,
	keywords_,
	constant_,
	identfier_,
	operator_tag_,
	delimiter_,
	delim_,
	whitespace_
};

regex digit_regex(digit);
regex keywords_regex(keywords);
regex constant_regex(constant);
regex identfier_regex(identfier);
regex operator_tag_regex(operator_tag);
regex delimiter_regex(delimiter);
regex whitespace_regex(whitespace);


bool check_is_separator(char);

bool get_word(stringstream& ss, string& str)
{
	string tmp;
	const static string separator = operator_tag + delimiter;
	if(not  ss << tmp)
	{
		return false;
	}
	if (regex_search(identfier, )  tmp.find_first_of(separator) not_eq string::npos)
	{
		return false;
	}
	
}

int main(int argc, char* argv[], char* env[])
{
	stringstream ss(cin);
	string current_word;
	while ()
	{
		if ()
	}

}

File: pl0_scanner.py
#! /usr/bin/python3

"""
    python 版本的词法扫描器
    其实将问题抽象起来就是从一个字符串流中提取每个符号
    将其做上标记并输出的过程
"""

import sys
import re


# 规则声明
digit        = r"[0-9]"

keywords     = r"const|var|procedure|begin|end|if|then|while|do|read|call|write|writeln|module|int"

constant     = r"[0-9]+"

identfier     = r"[A-Za-z_][A-Za-z_0-9]*"

operator     = r"\+|-|\*|\/|#|=|<|>|<>|:=|>=|<="

delimiter    = r"[,.;():]"

delim	     = r"\s"

whitespace   = r"\s+"

separator    = r"([+\-*/#,.;():\s]|<=|>=|:=|=|<|>)"

with open(sys.argv[1], mode='r') as pl0_file:
# 按行获取 pl0 源码， 每次的词法分析只处理一行
    content = []
    result = []
    content = pl0_file.readlines()
    for line in content:
        line = re.split(separator, line)
        line = [c for c in line if re.match('\S', c)]
        for word in line:
            if (re.match(keywords, word)):
                result.append((word, 1))
                continue
            if (re.match(operator, word)):
                result.append((word, 2))
                continue
            if (re.match(delimiter, word)):
                result.append((word, 3))
                continue
            if (re.match(constant, word)):
                result.append((word, 4))
                continue
            if (re.match(identfier, word)):
                result.append((word, 5))
                continue

            
        print(line)
    print(result)
